Sources/MapboxMaps/Annotations/AnnotationImagesManager.swift

```diff
@@ -0,0 +1,89 @@
+    func register(imagesConsumer: AnnotationImagesConsumer)
+    func unregister(imagesConsumer: AnnotationImagesConsumer)
+}
+
+@objc internal protocol AnnotationImagesConsumer {
+    func isUsingStyleImage(_ imageName: String) -> Bool
+}
+
+extension PointAnnotation {
+    public struct Image: Hashable {
+        public var image: UIImage
+        public var name: String
+
+        public init(image: UIImage, name: String) {
+            self.image = image
+            self.name = name
+        }
+    }
+}
+
+extension PointAnnotationManager: AnnotationImagesConsumer {
+
+    func isUsingStyleImage(_ imageName: String) -> Bool {
+        allImages.contains(imageName)
+    }
+}
+
+internal final class AnnotationImagesManager: AnnotationImagesManagerProtocol {
+
+    private let style: StyleProtocol
+    private var imagesConsumers = NSHashTable<AnnotationImagesConsumer>.weakObjects()
+    private var addedAnnotationImages = Set<String>()
+
+    init(style: StyleProtocol) {
+        self.style = style
+    }
+
+    /// Registers an ``AnnotationImagesConsumer``.
+    /// Upon removing an image from ``Style``, this ``AnnotationImagesManager`` will first consult
+    /// its consumers and only remove the image if there is no  consumers using this image.
+    func register(imagesConsumer: AnnotationImagesConsumer) {
+        imagesConsumers.add(imagesConsumer)
+    }
+
+    /// Unregisters an ``AnnotationImagesConsumer``.
+    func unregister(imagesConsumer: AnnotationImagesConsumer) {
+        imagesConsumers.remove(imagesConsumer)
+    }
+
+    /// Adds an image to ``Style`` if no image exists under same `id`.
+    func addImage(_ image: UIImage, id: String, sdf: Bool, contentInsets: UIEdgeInsets) {
+        guard !style.imageExists(withId: id) else { return }
+
+        do {
+            try style.addImage(image, id: id, sdf: sdf, contentInsets: contentInsets)
+            addedAnnotationImages.insert(id)
+        } catch {
+            Log.warning(
+                forMessage: "Could not add image to style due to error: \(error)",
+                category: "Annnotations")
+        }
+    }
+
+    /// Removes the image from ``Style`` if the image has been added/owned by this ``AnnotationImagesManager``
+    /// and if there is no consumers using this image.
+    func removeImage(_ imageName: String) {
+        guard canRemoveStyleImage(imageName) else { return }
+
+        do {
+            try style.removeImage(withId: imageName)
+            addedAnnotationImages.remove(imageName)
+        } catch {
+            Log.warning(
+                forMessage: "Could not remove image from style due to error: \(error)",
+                category: "Annnotations")
+        }
+    }
+
+    private func canRemoveStyleImage(_ imageName: String) -> Bool {
+        addedAnnotationImages.contains(imageName)
+        && imagesConsumers.allObjects.allSatisfy { !$0.isUsingStyleImage(imageName) }
+        && style.imageExists(withId: imageName)
+    }
+}
```

Comment:
I think image tracking can be sealed inside of the manager itself, this way its users won't need to store the ids of the images they are interested in.

```suggestion
internal final class AnnotationImagesManager: AnnotationImagesManagerProtocol {

    private let style: StyleProtocol
    private var imageReferenceCount = NSCountedSet()

    init(style: StyleProtocol) {
        self.style = style
    }

    func addImage(_ image: UIImage, id: String, sdf: Bool, contentInsets: UIEdgeInsets) {
        imageReferenceCount.add(id)

        guard !style.imageExists(withId: id) else { return }

        do {
            try style.addImage(image, id: id, sdf: sdf, contentInsets: contentInsets)
        } catch {
            Log.warning(
                forMessage: "Could not add image to style due to error: \(error)",
                category: "Annnotations")
        }
    }

    func removeImage(_ imageId: String) {
        defer { imageReferenceCount.remove(imageId) }

        guard imageReferenceCount.count(for: imageId) <= 1 else {
            return
        }

        do {
            try style.removeImage(withId: imageId)
        } catch {
            Log.warning(
                forMessage: "Could not remove image from style due to error: \(error)",
                category: "Annnotations")
        }
    }
}
```

Reply:
while the implementation using `NSCountedSet` can takeaway the image ids tracking from the consumers, it does not take into account when an image is added from an external source, in that case, that image is claimed by our internal consumers, and will be removed when our consumers release it. We don't want this behavior, if an image is added from an external source, we will not take ownership of it and thus, we should not remove it unless being told by that source.
Furthermore, I'd prefer the current approach, as it can contain information of who is using the image, imo this information can be helpful for debugging.
I do like the idea of extending this into other domain like Puck2D, I will take a look in this :-D